// cmp-plugin-base/dist/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cmp_baseconfig_1 = __importDefault(require("cmp-baseconfig"));
const cmp_logger_1 = __importDefault(require("cmp-logger"));
const lodash_isboolean_1 = __importDefault(require("lodash.isboolean"));
const package_json_1 = __importDefault(require("../package.json"));
class BasePlugin {
    constructor(pkg, options) {
        this.enabled = true;
        this.name = pkg.name.replace(/^cmp-plugin-/, '');
        this.version = pkg.version;
        this.description = pkg.description;
        this.baseVersion = package_json_1.default.version;
        this.options = options;
        this.baseConfig = cmp_baseconfig_1.default;
        this.logger = cmp_logger_1.default;
        if (this.init) {
            this.init();
        }
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async shouldRun(ctx) {
        if (this.options && this.options.enabled !== undefined) {
            if (lodash_isboolean_1.default(this.options.enabled)) {
                return this.options.enabled;
            }
            else {
                return (this.options.enabled.toLowerCase() === 'true');
            }
        }
        return true;
    }
}
exports.default = BasePlugin;
