// sso-olr/index.js
'use strict';

var sso        = require('./lib/sso'),
    olr        = require('./lib/olr'),
    partner    = require('./lib/partner'),
    middleware = require('./lib/middleware');

module.exports = function (endpoint) {
    var ssoClient = sso(endpoint),
        olrClient = olr(endpoint),
        partnerClient = partner(endpoint);

    return {
        getToken: ssoClient.getToken,
        getUser: ssoClient.getUser,
        getUserWithGroups: ssoClient.getUserWithGroups,
        refreshToken: ssoClient.refreshToken,
        getEntitlements: olrClient.getEntitlements,
        getEntitlementsForSection: olrClient.getEntitlementsForSection,
        getAllEntitlements: olrClient.getAllEntitlements,
        getCourse: olrClient.getCourse,
        getEnrollments: olrClient.getCourse,
        getEnrollmentsWithCourseInfo: olrClient.getEnrollmentsWithCourseInfo,
        partner: {
            getCourse: partnerClient.getCourse    
        },
        middleware: middleware(ssoClient)
    };
};
