// sso-olr/lib/olr.js
'use strict';

var ssoOlrRequest = require('./request');

function olr(endpoint) {
    var getEndpoint = endpoint;
    if (endpoint !== undefined && typeof endpoint !== 'function') {
        getEndpoint = function() {
            return endpoint;
        };
    }

    function getAllEntitlements(token, callback) {
        ssoOlrRequest(getEndpoint() + '/olrws/rest/getEntitlements', {
            userToken: token
        }, callback);
    }

    function getEnrollments(token, guid, callback) {
        ssoOlrRequest(getEndpoint() + '/olrws/rest/getEnrollments', {
            token: token,
            guid: guid
        }, callback);
    }


    function getEnrollmentsWithCourseInfo(token, guid, callback) {
        ssoOlrRequest(getEndpoint() + '/olrws/rest/getEnrollments', {
            token: token,
            guid: guid,
            requestCourseInfo: true
        }, callback);
    }

    function getCourse(token, courseKey, callback) {
        ssoOlrRequest(getEndpoint() + '/olrws/rest/getCourse', {
            token: token,
            courseKey: courseKey
        }, callback);
    }

    function getEntitlements(token, eisbn, callback) {
        ssoOlrRequest(getEndpoint() + '/olrws/rest/getEntitlements', {
            userToken: token,
            filter : {
                filterType : 'EISBN',
                filerValue : eisbn
            }
        }, callback);
    }

    return {
        getEntitlements: getEntitlements,
        getAllEntitlements: getAllEntitlements,
        getCourse: getCourse,
        getEnrollments: getEnrollments,
        getEnrollmentsWithCourseInfo: getEnrollmentsWithCourseInfo
    };
}

module.exports = olr;
