// cmp-baseconfig/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rc_1 = __importDefault(require("rc"));
const path_1 = require("path");
const http_1 = require("http");
exports.default = rc_1.default('cmp', {
    name: process.env.CMP_PACKAGE_TITLE || process.env.npm_package_name,
    version: process.env.CMP_PACKAGE_VERSION || process.env.npm_package_version,
    engines: {
        node: process.version
    },
    port: 8080,
    agent: {
        maxSockets: Infinity
    },
    session: {
        cookie: {
            httpOnly: true,
            sameSite: 'lax'
        }
    },
    http: {
        timeout: 120000,
        keepAliveTimeout: 65000,
        headersTimeout: 65000,
        maxHeadersCount: 2000,
        maxHeaderSize: http_1.maxHeaderSize // Set by CLI option
    },
    dnsCache: {
        enable: true,
        family: 4,
        all: false
    },
    crowdControl: false,
    policies: {
        ttl: 1000 * 60 * 60,
        pollInterval: 1000 * 60 * 5,
        directory: path_1.join(process.cwd(), 'policies') /*,
        ces: {
          url: 'https://ces-awsbase.cengage.info/services/dataservice',
          namespace: 'cmp',
          environment: 'localhost',
          ceq: {
            hosts: [
              'ceq-awsbase01.cengage.info',
              'ceq-awsbase02.cengage.info'
            ],
            autoAcknowledge: true
          }
        }*/
    },
    plugins: {
        directory: path_1.join(process.cwd(), 'plugins', (process.env.NODE_APP_INSTANCE) ? process.env.NODE_APP_INSTANCE : '')
    },
    ssl: {
        port: 8081,
        cert: '',
        key: ''
    },
    logging: {
        target: 'console',
        level: 'debug',
        console: {
            pretty: true
        }
    }
});
