// sso-olr/lib/sso.js
'use strict';

var ssoOlrRequest = require('./request');

function sso(endpoint) {
    var getEndpoint = endpoint;
    if (endpoint !== undefined && typeof endpoint !== 'function') {
        getEndpoint = function() {
            return endpoint;
        };
    }

    function getToken(username, password, callback) {
        ssoOlrRequest(getEndpoint() + '/ssows/rest/getToken', {
            uid: username,
            password: password
        }, callback);
    }

    function refreshToken(token, callback) {
        ssoOlrRequest(getEndpoint() + '/ssows/rest/getToken', {
            userToken: token,
            adminToken: token
        }, callback);
    }

    function getUser(token, callback) {
        ssoOlrRequest(getEndpoint() + '/ssows/rest/getUser', {
            token: token,
            returnGroups: false
        }, callback);
    }

    function getUserWithGroups(token, callback) {
        ssoOlrRequest(getEndpoint() + '/ssows/rest/getUser', {
            token: token,
            returnGroups: true
        }, callback);
    }

    return {
        getToken: getToken,
        getUser: getUser,
        getUserWithGroups: getUserWithGroups,
        refreshToken: refreshToken
    };
}

module.exports = sso;
